packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -3,30 +3,152 @@
         wPowerPerpPool = _wPowerPerpPool;
+        wPowerPerp = _wPowerPerp;
+        weth = _weth;
+    }
+
+    /**
+     * @dev accept erc721 from safeTransferFrom and safeMint after callback
+     * @return returns received selector
+     */
+    function onERC721Received(
+        address,
+        address,
+        uint256,
+        bytes memory
+    ) public virtual override returns (bytes4) {
+        return this.onERC721Received.selector;
+    }
+
+    /**
+     * @notice receive function to allow ETH transfer to this contract
+     */
+    receive() external payable {
+        require(msg.sender == weth || msg.sender == address(controller), "Cannot receive eth");
+    }
+
+    function flashswapWMint(
+        uint256 _vaultId,
+        uint256 _wPowerPerpAmount,
+        uint256 _collateralAmount
+    ) external payable {
+        uint256 amountToFlashswap = _collateralAmount.sub(msg.value);
+
+        _exactInFlashSwap(
+            wPowerPerp,
+            weth,
+            IUniswapV3Pool(wPowerPerpPool).fee(),
+            _wPowerPerpAmount,
+            amountToFlashswap,
+            uint8(FLASH_SOURCE.FLASH_W_MINT),
+            abi.encodePacked(_vaultId, amountToFlashswap, _collateralAmount, _wPowerPerpAmount)
+        );
+
+        emit FlashswapWMint(msg.sender, _vaultId, _wPowerPerpAmount, amountToFlashswap, _collateralAmount);
+    }
+
+    /**
+     * @notice uniswap flash swap callback function
+     * @dev this function will be called by flashswap callback function uniswapV3SwapCallback()
+     * @param _caller address of original function caller
+     * @param _amountToPay amount to pay back for flashswap
+     * @param _callData arbitrary data attached to callback
+     * @param _callSource identifier for which function triggered callback
+     */
+    function _swapCallback(
+        address _caller,
+        address, /*_tokenIn*/
+        address, /*_tokenOut*/
+        uint24, /*_fee*/
+        uint256 _amountToPay,
+        bytes memory _callData,
+        uint8 _callSource
+    ) internal override {
+        if (FLASH_SOURCE(_callSource) == FLASH_SOURCE.FLASH_W_MINT) {
+            flashswapWMintData memory data = abi.decode(_callData, (flashswapWMintData));
+
+            // convert WETH to ETH as Uniswap uses WETH
+            IWETH9(weth).withdraw(IWETH9(weth).balanceOf(address(this)));
+
+            //will revert if data.flashSwappedCollateral is > eth balance in contract
+            // IController(controller).mintWPowerPerpAmount{value: address(this).balance}(data.vaultId, data.wPowerPerpAmount, 0);
+            uint256 vaultId = IController(controller).mintWPowerPerpAmount{value: data.totalCollateralToDeposit}(
+                data.vaultId,
+                data.wPowerPerpAmount,
+                0
+            );
+
+            //repay the flash swap
+            IWPowerPerp(wPowerPerp).transfer(wPowerPerpPool, _amountToPay);
+
+            if (address(this).balance > 0) {
```

Comment:
what happens if we try to send 0 ETH?

for pretty much every case there will be some remaining eth in the contract (at least 1 wei), so is it cheaper to just send address(this).balance to the user ever time and not check, or will it revert?